### DISCUSSION TITLE
マルチパッケージ構成でナビゲーション、ルーターを“循環参照しないように“使用する方法を考える

### DISCUSSION THREAD

**Discussion by Yamasaki-pan961** [yumemi-inc/flutter-mobile-project-template#76] [General]
- https://github.com/yumemi-inc/flutter-mobile-project-template/pull/73

マルチパッケージ構成でナビゲーション、ルーターを“循環参照しないように“使用する方法はいくつかあり、どれを採用するか決めるのは難しいと思い、、、
一旦メリット・デメリット挙げてみました。

1. 各パッケージでそれぞれナビゲーション処理を呼び出したり、ルートを公開する（シングルパッケージ構成と同じ）
[メリット]
考えることが少ないので楽
[デメリット]
全てのパッケージが特定のナビゲーションパッケージに依存することになるので、パッケージ変更の際にコストがかかる

2. 各パッケージでナビゲーション処理を受け取る引数を定義し、appから処理を渡す
[メリット]
appのみナビゲーションパッケージに依存するためパッケージの変更に強い
[デメリット]
各featureパッケージでナビゲーション処理を受け取れる引数を定義する必要があり、遷移が多いと非常に面倒になる

3. coreパッケージでナビゲーション処理のインターフェースを定義し、各パッケージで呼び出す
[メリット]
1と2のデメリットを解消できる
[デメリット]
GoRouterのような高機能なパッケージのインターフェースを定義するのは大変

**Comment by Yamasaki-pan961** [comment]
# 問題の切り分け
## 1. 依存関係の制約
以下の制約を守る必要がある。
### 制約
```
可能
app -> cores
app -> feature
feature -> cores

不可 
cores -> feature
cores -> app
feature: A -> feature: B
```

仮に`feature: A`のWidgetから`feature: B`の画面に遷移したい場合、必ず`2`の方法をとらなければ制約を守れない。
> 2. 各パッケージでナビゲーション処理を受け取る引数を定義し、appから処理を渡す


## 2. 遷移が多い場合、`2`の方法だと引数が多くなる
解決策としては以下が思いつく
> 同じfeature内にある画面のルートクラスは、同じfeatureパッケージ内に記述し呼び出す

これの問題点は、
[a] 各featureがそれぞれGoRouterなどのパッケージに依存するようになる
[b] ルートクラスの定義がfeatureとappで混同し運用が複雑になる
`a`は`3`の方法で解決できるが`b`が解決できない。
> 3. coreパッケージでナビゲーション処理のインターフェースを定義し、各パッケージで呼び出す

また`3`の方法のデメリットがある
> GoRouterのような高機能なパッケージのインターフェースを定義するのは大変

**Comment by morikann** [comment]
### 結論
2 で進めていく。

### 今後の方針
現状、router_config.dart に全ての route が定義されているが、ページが増えるにつれて router_config.dart が肥大化してしまう恐れがある。

そのため、プロジェクト毎に routes 以下に適切に route ファイルを配置する。
```
└── router
    ├── routes
    │   ├── page_1_route.dart
    │   └── page_2_route.dart
    └── provider
        ├── router.dart
        └── router.g.dart
```

以下は参考例です。

```dart
@TypedGoRoute<HomePageRoute>(
    path: HomePageRoute.path,
    routes: [
      TypedGoRoute<SettingRoute>(
        path: SettingRoute.path,
      )
    ]
)

class HomePageRoute extends GoRouteData {
  const HomePageRoute();

  static const path = '/home';

  @override
  Widget build(BuildContext context, GoRouterState state) {
    return const HomePage(
      onSettingIconPressed: SettingRoute().go(context);
    );
  }
}

class SettingRoute extends GoRouteData {
  const SettingRoute();

  static const path = 'setting';

  @override
  Widget build(BuildContext context, GoRouterState state) {
    return SettingPage();
  }
}
```

**Reply by Kotaro666-dev** [reply]
@morikann 

本ディスカッションはこちらのコメントで結論が出されたと認識しています。
その一方で、本ディスカッションが引き続き open となっている背景はありますか？ 👀 

もし本ディスカッションを open 状態で保持して議論を続ける必要がなかれば、 close したほうが良いと思いました！
